rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/FileMetadata.java

```diff
@@ -55,15 +58,29 @@ public FileMetadata()
   {
   }
 
-  public FileMetadata(String remotePath, long fileLength, long lastModified, long currentFileSize, Configuration conf)
+  public FileMetadata(String remotePath, long fileLength, long lastModified, long currentFileSize, Configuration conf, boolean updateMetadata)
       throws IOException
   {
     this.remotePath = remotePath;
     this.size = fileLength;
     this.lastModified = lastModified;
     this.currentFileSize = currentFileSize;
-    localPath = CacheUtil.getLocalPath(remotePath, conf);
-    mdFilePath = CacheUtil.getMetadataFilePath(remotePath, conf);
+    if (!updateMetadata) {
+      if (generationNumber.containsKey(remotePath)) {
+        genNumber = generationNumber.get(remotePath);
+        genNumber++;
```

Comment:
This block is not thread safe. Multiple getCacheStatus on file  can come in parallel. 
Make sure when you make this thread safe, you dont block the requests on other file on the exists() check later.